Implement ProtocolFeatureStore

dkropachev opened the pull request:
Implement class to store information about all features negotiated.
The main idea is to have a single place where we store that information, instead of spreading all over channel, and `ProtocolInitHandler`

It is required by https://github.com/scylladb/java-driver/pull/599

dkropachev committed:
Implement ProtocolFeatureStore

Implement class to store information about all features negotiated.

dkropachev committed:
Pull ShardingInfo into ProtocolFeatureStore

dkropachev committed:
Pull TabletInfo into ProtocolFeatureStore

Bouncheck reviewed (approved):
LGTM.

There may be very slight performance hit, probably negligible.
Instead of getting specific features directly the getters will now get whole feature store and then the specific feature from it.

Bouncheck commented:
I see that two features had method names changed to `populateStartupOptions` and `loadFromSupportedOptions`. If you'd like to enforce a convention for the future features consider making an `Interface` and turning the existing ones into implementation of it

dkropachev committed:
Add ProtocolFeatureStore cache in DriverChannel

No need to pull this information all the time from the channel.
Let's cache it when it is populated to the channel and use it.

dkropachev commented:
> I see that two features had method names changed to `populateStartupOptions` and `loadFromSupportedOptions`. If you'd like to enforce a convention for the future features consider making an `Interface` and turning the existing ones into implementation of it

I was thinking about the same, `loadFromSupportedOptions` is static, so it can't go into interface and having interface for only one method does not make sense, so I did not do it.
